### ISSUE TITLE
line 149 training_step관련 질의사항

### ISSUE THREAD

**Issue by Ethan0625** [haven-jeon/KoGPT2-chatbot#17]
안녕하세요 희원님. training step의 loss_avg관련하여 질의사항이 있어 issue를 남깁니다.

train_torch.py의 line 155에 있는 loss_avg부분을 보시면

loss_avg = loss.sum() / mask.sum()으로 되어있습니다.

loss.sum()의 경우는 Q와 A로 인한 loss를 나타내는 tensor값들의 합이고

mask.sum()의 경우는 line 70의 __getitem__에서 만들어진 mask의 합으로 알고있습니다.

그렇기때문에 mask.sum()의 경우에는 실제 loss.mean()보다 값이 크게 나오는데

혹시 의도하신 부분인건지 궁금합니다.

**Comment by haven-jeon** [comment]
이렇게 계산하고자 한 목적은 Answer에 해당하는 텐서로만 loss를 계산하기 위함입니다. 

코드 보시면 아시겠지만 loss는 `reduction='none'` 상태로 리턴이 됩니다. 따라서 loss는 (batchsize, seq_len), mask는 (batchsize, seq_len)이 되겠죠. 여기서 mask.sum()으로 나눠주지 않고  loss.mean() 만을 사용하게 되면 Answer에 해당하는 시퀀스 부분 이외의 길이까지 고려하기 때문에 loss가 실제보다 과도하게 적게 나와 최적화에 문제가 생길겁니다. 

인코더 디코더의 경우 이렇게 복잡하지 않은데, 입력 시퀀스 길이가 매번 달라지니 마스킹을 복잡하게 적용하게 되네요..  ^^;

질문 감사합니다...

**Comment by haven-jeon** [comment]
별도 피드백이 없어 해당 이슈 클로즈 합니다.
